jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/RetainingResponseListener.java

```diff
@@ -0,0 +1,111 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.client;
+
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+
+import org.eclipse.jetty.io.Content;
+import org.eclipse.jetty.io.RetainableByteBuffer;
+import org.eclipse.jetty.util.BufferUtil;
+
+/**
+ * <p>Implementation of {@link Response.Listener} that retains response content up to a configurable number of bytes.</p>
+ * <p>The content may be retrieved from {@link #onSuccess(Response)} or {@link #onComplete(Result)}
+ * via {@link #getContent()} or {@link #getContentAsString()}.</p>
+ * <p>Instances of this class are not reusable, so one must be allocated for each request.</p>
+ * <p>The implementation retains the response content chunks, and converts them into a {@code byte[]}
+ * upon response success, returned by {@link #getContent()}.</p>
+ */
+public abstract class RetainingResponseListener extends AbstractResponseListener
+{
+    private final RetainableByteBuffer.DynamicCapacity accumulator;
+    private byte[] content;
+
+    public RetainingResponseListener()
+    {
+        this(2 * 1024 * 1024);
+    }
+
+    public RetainingResponseListener(int maxLength)
+    {
+        super(maxLength);
+        this.accumulator = new RetainableByteBuffer.DynamicCapacity(null, maxLength, 0);
+    }
+
+    @Override
+    public void onContent(Response response, Content.Chunk chunk, Runnable demander) throws Exception
+    {
+        chunk.retain();
+        accumulator.add(chunk);
```

Comment:
It would be better to use the pattern below, so that the RBB can be configured (or heuristically) decide if a copy or a retain is better.   This class might then be better named `AccumulatingResponseListener`.

```suggestion
        accumulator.append(chunk);
```

Reply:
I don't want to copy, I want exactly retain+add to a list.
And I don't want any heuristic to decide for me.

I already have the chunk, there is no need to copy, by allocating _another_ buffer, likely of the wrong capacity, to hold copied bytes that don't need to be copied.
And that would require a `ByteBufferPool` or otherwise be extremely inefficient by allocating buffers to copy into that are not pooled.

I only use `DynamicCapacity` for `takeByteArray()`, but I really just need a `List`, which is why `DynamicCapacity` was configured to not do any heuristic.

Reply:
How you know that you don't want to copy?    If you are given 100 x 1 bytes chunks, each coming in 32KB buffer, then it is probably better to copy rather than retain.   Or retain the first and then use its space to aggregate others (in some circumstances).

The whole point of RBB is that it removes the need to codify such behaviours in multiple places all over the code base whenever you interact with the RBB.  Instead, the behaviour can be specified once at the time of construction and then all interactions with the RBB can use simple methods that will work with whatever algorithm is used.

If you want the default to always retain, then construct the RBB in the default constructor to do so.  You then don't need to hard code the algorithm here and it can easily be changed by simply using a different RBB.

Note also, that if RBB needs a pool, it can discover one from the passed RBB if it is a pooled one already.

In short, there are many optimizations/smarts built into RBB that might be of use and there is no reason to bypass them and assume specific behaviour in multiple places in your code. If you want a specific algorithm, then specify that in the constructor that constructs the RBB, not in each method that interacts with the RBB.

In fact, rather than having `RetainingResponseListener` and `BufferingResponseListener` as two different classes, they should just be the one class with two different constructors, as all the difference in behaviour can be encoded into the RBB: always copy; always retain; or heuristically decide between the two options.  Using RBB properly will avoid the need for a whole new class!